chaincode.go

```diff
@@ -0,0 +1,377 @@
+	err := json.Unmarshal([]byte(args[0]), &record)
+
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	assetBytes, _ := stub.GetState(record.AssetId)
+
+	if len(assetBytes) == 0 {
+		return shim.Error(fmt.Sprintf(constants.ErrorAssetIdNotFound, record.AssetId))
+	}
+
+	recordId, err := utils.CreateCompositeKey(stub, constants.PrefixRecord)
+
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	record.Id = recordId
+	record.DateCreated = time.Now()
+ 
+	jsonRecord, _ := json.Marshal(record)
+
+	err = stub.PutState(record.Id, jsonRecord)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	return shim.Success(jsonRecord)
+}
+
+func (scc *SupplyChainChaincode) placeOrder(stub shim.ChaincodeStubInterface, args []string) peer.Response {
+	if len(args) != 1 {
+		return shim.Error(constants.ErrorArgumentsLength)
+	}
+
+	order := types.Order {}
+	err := json.Unmarshal([]byte(args[0]), &order)
+
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	orderId, err := utils.CreateCompositeKey(stub, constants.PrefixOrder)
+
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	order.Id = orderId
+	order.DateCreated = time.Now()
+	order.IsCompleted = false
+
+	jsonOrder, _ := json.Marshal(order)
+
+	err = stub.PutState(order.Id, jsonOrder)
+
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	return shim.Success(jsonOrder)
+}
+
+func (scc *SupplyChainChaincode) fulfillOrder(stub shim.ChaincodeStubInterface, args []string) peer.Response {
+	if len(args) != 1 {
+		return shim.Error(constants.ErrorArgumentsLength)
+	}
+
+	orderFulfillment := types.OrderFulfillment {}
+	err := json.Unmarshal([]byte(args[0]), &orderFulfillment)
+
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	orderBytes, _ := stub.GetState(orderFulfillment.Id)
+
+	if len(orderBytes) == 0 {
+		return shim.Error(fmt.Sprintf(constants.ErrorOrderIdNotFound, orderFulfillment.Id))
+	}
+
+	order := types.Order {}
+	err = json.Unmarshal(orderBytes, &order)
+
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	if order.IsCompleted == true {
+		return shim.Error(fmt.Sprintf(constants.ErrorOrderIsFulfilled,orderFulfillment.Id))
+	}
+
+	for _, recordElem := range orderFulfillment.Records {
+		recordBytes, _ := stub.GetState(recordElem.Id)
+
+		if len(recordBytes) == 0 {
+			return shim.Error(fmt.Sprintf(constants.ErrorRecordIdNotFound, recordElem.Id))
+		}
+		record := types.Record {}
+		err := json.Unmarshal(recordBytes, &record)
+
+		if err != nil {
+			return shim.Error(err.Error())
+		}
+
+		if record.AssetId != order.AssetId {
+			return shim.Error(fmt.Sprintf(constants.ErrorRecordDifferentAssetId, record.Id, record.AssetId, order.AssetId))
+		}
+
+		if recordElem.Quantity > record.Quantity {
+			return shim.Error(fmt.Sprintf(constants.ErrorRecordQuantity, recordElem.Id))
+		}
+		
+		record.Quantity -= recordElem.Quantity
+		record.LastUpdated = time.Now()
+
+		newRecord := types.Record {
+			AssetId: record.AssetId,
```

Comment:
Looking at it, I am wondering whether we should also have a param field "creationOrderId" in order for us to be able to find all records created via fulfilment of a certain order